mono/metadata/class-init.c

```diff
@@ -3846,40 +3844,36 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 			real_size = MAX (real_size, size + field_offsets [i]);
 		}
 
-		if (klass->has_references) {
-			ref_bitmap = g_new0 (guint8, real_size / TARGET_SIZEOF_VOID_P);
-
-			/* Check for overlapping reference and non-reference fields */
-			for (i = 0; i < top; i++) {
-				MonoType *ftype;
-
+		/* check for incorrectly aligned or overlapped by a non-object field */
+#ifdef ENABLE_NETCORE	
+		guint8 *layout_check;	
+		if (has_references) {
+			layout_check = g_new0 (guint8, real_size);
+			for (i = 0; i < top && !mono_class_has_failure (klass); i++) {
 				field = &klass->fields [i];
-
-				if (mono_field_is_deleted (field))
-					continue;
-				if (field->type->attrs & FIELD_ATTRIBUTE_STATIC)
+				if (!field)
 					continue;
-				ftype = mono_type_get_underlying_type (field->type);
-				if (MONO_TYPE_IS_REFERENCE (ftype))
-					ref_bitmap [field_offsets [i] / TARGET_SIZEOF_VOID_P] = 1;
-			}
-			for (i = 0; i < top; i++) {
-				field = &klass->fields [i];
-
 				if (mono_field_is_deleted (field))
 					continue;
 				if (field->type->attrs & FIELD_ATTRIBUTE_STATIC)
 					continue;
-
-				// FIXME: Too much code does this
-#if 0
-				if (!MONO_TYPE_IS_REFERENCE (field->type) && ref_bitmap [field_offsets [i] / TARGET_SIZEOF_VOID_P]) {
-					mono_class_set_type_load_failure (klass, "Could not load type '%s' because it contains an object field at offset %d that is incorrectly aligned or overlapped by a non-object field.", klass->name, field_offsets [i]);
+				int align = 0;
+				int size = mono_type_size (field->type, &align);
+				int j = 0;
+				MonoType *ftype = mono_type_get_underlying_type (field->type);
+				ftype = mono_type_get_basic_type_from_generic (ftype);
+				guint8 type =  type_has_references (klass, ftype) ? 1 : 2;				
+				for (j = 0; j < size; j++) {
+					if (layout_check[field_offsets [i] + j] != 0 && layout_check[field_offsets [i] + j] != type) {
```

Comment:
This could be for (int j = 0;

Reply:
Needs more spaces to conform to coding style.

Reply:
Thanks!